src/actor.rs

```diff
@@ -0,0 +1,196 @@
+use futures::prelude::*;
+use futures::stream::FuturesUnordered;
+use ractor::{
+    concurrency::Duration, Actor, ActorProcessingErr, ActorRef, BytesConvertable, RpcReplyPort,
+};
+use ractor_cluster::RactorClusterMessage;
+use serde::{Deserialize, Serialize};
+use threshold_crypto::{PublicKeySet, SecretKeyShare, Signature, SignatureShare};
+
+const MPC_RECOVERY_GROUP: &str = "mpc-recovery";
+
+type NodeId = u64;
+type Payload = Vec<u8>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignResponse {
+    node_id: NodeId,
+    sig_share: SignatureShare,
+}
+
+impl BytesConvertable for SignResponse {
+    fn into_bytes(self) -> Vec<u8> {
+        serde_json::to_vec(&self).unwrap()
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        serde_json::from_slice(&bytes).unwrap()
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignatureResponse {
+    pub sig: Signature,
+}
+
+impl BytesConvertable for SignatureResponse {
+    fn into_bytes(self) -> Vec<u8> {
+        serde_json::to_vec(&self).unwrap()
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        serde_json::from_slice(&bytes).unwrap()
+    }
+}
+
+#[derive(RactorClusterMessage, Debug)]
+pub enum NodeMessage {
+    #[rpc]
+    NewRequest(Payload, RpcReplyPort<SignatureResponse>),
+    #[rpc]
+    SignRequest(Payload, RpcReplyPort<SignResponse>),
+}
+
+pub struct NodeActor;
+
+pub struct NodeActorState {
+    id: NodeId,
+    pk_set: PublicKeySet,
+    sk_share: SecretKeyShare,
+}
+
+#[async_trait::async_trait]
+impl Actor for NodeActor {
+    // An actor has a message type
+    type Msg = NodeMessage;
+    // and (optionally) internal state
+    type State = NodeActorState;
+    // Startup initialization args
+    type Arguments = (NodeId, PublicKeySet, SecretKeyShare);
+
+    async fn pre_start(
+        &self,
+        myself: ActorRef<Self>,
+        args: (NodeId, PublicKeySet, SecretKeyShare),
+    ) -> Result<Self::State, ActorProcessingErr> {
+        ractor::pg::join(MPC_RECOVERY_GROUP.to_string(), vec![myself.get_cell()]);
+        // create the initial state
+        Ok(NodeActorState {
+            id: args.0,
+            pk_set: args.1,
+            sk_share: args.2,
+        })
+    }
+
+    // This is our main message handler
+    async fn handle(
+        &self,
+        _myself: ActorRef<Self>,
+        message: Self::Msg,
+        state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        let remote_actors = ractor::pg::get_members(&MPC_RECOVERY_GROUP.to_string())
```

Comment:
Do we need to fetch all the actors in each `handle()` execution?

Reply:
If we want the actor system to be dynamically joinable, then yes. But as I said we could also just get rid of actors entirely and do `N` HTTP requests to static addresses instead.

Reply:
It can be dinamically joinable in the future, but I believe it's fixed for now.
Ok, let's keep it this way. I guess there is no harm from this design at the initial stage.